Add multitenant support

sileht opened the pull request:
Add multitenant
tests: cleanup subscriptions at startup

sileht commented:
All tests in:
* test_subscriptions.py
* test_tenant_migration.py
* test_http_callback.py

fblackburn1 commented on integration_tests/suite/test_subscriptions.py:
```diff
@@ -225,14 +223,19 @@ def test_when_create_invalid_subscription_then_400(self):
 
     def test_when_create_http_subscription_then_subscription_no_error(self):
         webhookd = self.make_webhookd(VALID_TOKEN)
+        tenant_uuid = "ffffffff-ffff-ffff-ffff-ffffffffffff"
```
The tenant used to create the subscription is extracted from the token. So we should import information from the token used. But this token is hardcoded in wazo-auth-mock, se we use constant, and in this case we should import WAZO_TENANT from the base helper, not an arbitrary specific constant for this test.

sileht replied:
done

fblackburn1 commented:
> It fixes all tests that use `base` class and the `subscription` fixture:
> 
>     * test_subscriptions.py
> 
>     * test_tenant_migration.py
> 
>     * test_http_callback.py

Ah when using [denv](https://github.com/wazo-pbx/denv)? :stuck_out_tongue:

fblackburn1 commented on alembic/versions/f63e43dcf0ca_add_tenant_to_subscription.py:
```diff
@@ -0,0 +1,29 @@
+    op.alter_column('webhookd_subscription', 'tenant_uuid', server_default=None)
+
+def downgrade():
+    pass
```
We try to make downgrade when the impact is the schema and not the data. It's generally easy to write and easy for the reviewer :smile:

sileht replied:
done

sileht commented on wazo_webhookd/database/models.py:
```diff
@@ -112,6 +112,7 @@ def update(self, **attributes):
                            Column('uuid', String(38), server_default=text('uuid_generate_v4()'), primary_key=True),
                            Column('name', Text()),
                            Column('service', Text()),
+                           Column('tenant_uuid', String(36)),
```
done

sileht commented on wazo_webhookd/plugins/api/api.yml:
```diff
@@ -57,3 +57,16 @@ parameters:
+      type: string
+      in: header
+      description: "The tenant's UUID, defining the ownership of a given resource."
+      required: false
```
done

sileht commented on wazo_webhookd/plugins/subscription/resource.py:
```diff
@@ -58,36 +89,33 @@ def delete(self, subscription_uuid):
         subscription['events_user_uuid'] = subscription['owner_user_uuid'] = token.user_uuid()
         subscription['events_wazo_uuid'] = token.wazo_uuid()
         subscription['uuid'] = str(uuid.uuid4())
+        print("POST USER: %s" % Tenant.autodetect().uuid)
```
done

sileht commented on etc/wazo-webhookd/config.yml:
```diff
@@ -95,6 +95,7 @@ enabled_plugins:
   services: true
   status: true
   subscriptions: true
+  tenant_migration: false
```
done

sileht commented:
> > It fixes all tests that use `base` class and the `subscription` fixture:
> > ```
> > * test_subscriptions.py
> > 
> > * test_tenant_migration.py
> > 
> > * test_http_callback.py
> > ```
> 
> Ah when using [denv](https://github.com/wazo-pbx/denv)? stuck_out_tongue

Not only, when one test fails and other have good change to fail because the subscription db can not be clean.
But yeah with denv it have 100% change to occurs without this :)

sileht commented:
> Missing multi-tenant tests: there are no tests on unauthorized errors. You can check "multi_tenant" tests in https://github.com/wazo-pbx/xivo-confd/blob/master/integration_tests/suite/base/test_call_permissions.py

I will add them

> Missing swagger documentation on 401 error (Is not caused by this branch, but this branch add more 401 return code)

Done

> I can GET/DELETE webhookd if I know the uuid even if I'm not in an authorized tenant. You need to extract tenant to validate that the user can see the webhook.

Fixed,  will add test too.

> Hint to extract user_uuid AND tenant_uuid in the same request to have more symmetric (calling all service methods with tenant_uuids): https://github.com/wazo-pbx/wazo-chatd/blob/master/wazo_chatd/plugins/rooms/http.py#L82

Looks great I will do something similar. I have also added https://github.com/wazo-pbx/xivo-lib-python/pull/9 to also share visible_tenants(), instead of copying it.

fblackburn1 commented on integration_tests/suite/test_subscriptions.py:
```diff
@@ -217,24 +225,30 @@ def test_given_wrong_auth_when_create_subscription_then_401(self):
         subscription_uuid = response['uuid']
 
+        assert_that(response, has_key('owner_tenant_uuid'))
+        assert_that(response['owner_tenant_uuid'], MASTER_TENANT)
```
You can use `assert_that(response, has_entries(owner_tenant_uuid=MASTER_TENANT))` instead of checking the key and after checking the value

sileht replied:
done

fblackburn1 replied:
An now you can delete `assert_that(response, has_key('owner_tenant_uuid'))` :)

fblackburn1 commented on integration_tests/suite/test_subscriptions.py:
```diff
@@ -1,28 +1,37 @@
+    USERS_TENANT,
+    OTHER_USER_TOKEN,
+)
 from .helpers.fixtures import subscription
```
[Outside of scope (please do not change it in this PR)]
In the others repo we just `from .helpers import fixtures` and use `fixtures.subscription`. Since the name fixtures is not in the name of the object, is more appropriate. And we do not need to set some weird suffix (i.e. : `subscription_`)

sileht replied:
Completely agree

sileht commented on wazo_webhookd/plugins/tenant_migration/http.py:
```diff
@@ -0,0 +1,22 @@
+            self._service.update_owner_tenant_uuid(**item)
+        token = self._auth_client.token.new(expiration=60)
+        self._service.update_remaining_owner_tenant_uuid(
+            token['metadata']['tenant_uuid'])
```
done

sileht commented on wazo_webhookd/controller.py:
```diff
@@ -50,6 +54,7 @@ def __init__(self, config):
                 'api': api,
                 'bus_consumer': self._bus_consumer,
                 'config': config,
+                'auth_client': self._auth_client,
```
I have remove this.

It's currently passed to service plugins, too. I will do this change in an other PR

fblackburn1 replied:
Ah yeah I didn't see it. For http plugin, my main argument it's for homogeneity, but for service plugin, since only webhookd have it, I don't have any argument ... :sweat_smile:
